### ISSUE TITLE
[Feature request][FSDP] Automatically consider the module with "mixed `requires_grad`" as the ignored module.

### ISSUE THREAD

**Issue by HAOCHENYE** [pytorch/pytorch#104494]
### 🚀 The feature, motivation and pitch

@awgu 

Hi, as mentioned in issue #75943, it is common to train a model where some parameters require gradients while others do not. PR #98221 solves this problem when `use_orig_params` is set to `True`."

However, Most of the beginners of FSDP will still meet the problem since the `use_orig_params` defaults to `False`. 

BTW, is there any other drawbacks of setting `use_orig_params` to True? I've checked the issue #85831, and it seems most of the restrictions (<===) of  `use_orig_params=True` are corner cases for most of users.

if FSDP will keep `use_orig_params=False` as a default value. Will it be a good idea to automatically find the "gradient mixed" module and consider them as `ignored_modules`? I have written a draft as follows:

```python
        auto_ignored_modules = []

        # Here are three states for a module:
        # 1. `true` module: All parameters require the gradient
        # 2. `false` module: None of the parameters requrires the gradient
        # 3. `mixed` module: Only part of the parameters requrie the gradient

        def get_grad_state(module):
            if all(p.requires_grad for p in module.parameters()):
                return 'true'
            elif all(not p.requires_grad for p in module.parameters()):
                return 'false'
            else:
                return 'mixed'

        # Here are some rules to recognize the ignored modules:
        # 1. If `module._modules` is empty, such as `nn.Linear`, it will be
        #    recognized as an ignored module if its parameters is `mixed`
        # 2. `module._parameters` is empty, such as `ResNet50`, it will be
        #    recognized as an ignored module if both `true` module and `false`
        #    module exist in its submodules.
        # 3. For other cases, if its parameters are `mixed` or the state of
        #    its submodules diff from the state of the parameters, it will be
        #    recognized an ignored module.
        def find_ignored_modules(root_module: nn.Module):
            if not list(root_module.children()):
                if 'mixed' == get_grad_state(root_module):
                    auto_ignored_modules.append(root_module)
            else:
                children = list(root_module.children())
                params = list(root_module.parameters(recurse=False))
                if children and not params:
                    states = list(get_grad_state(m) for m in children)
                    if not (all(s == 'true' for s in states)
                            or all(s == 'false'
                                   for s in states) or 'mixed' in states):
                        auto_ignored_modules.append(root_module)
                        return
                elif params and children:
                    if all(p.requires_grad for p in params):
                        state = 'true'
                    elif all(not p.requires_grad for p in params):
                        state = 'false'
                    else:
                        auto_ignored_modules.append(root_module)
                        return
                    states = list(get_grad_state(m) for m in children)
                    if not all(s == state for s in states):
                        auto_ignored_modules.append(root_module)
                        return
                for child in children:
                    find_ignored_modules(child)
        
        # where module is the `root_module`
        find_ignored_modules(module)

        # If the state of the root module is mixed, but no ignored submodule
        # is found, the whole root module will not be sharded.
        if get_grad_state(module) == 'mixed' and not auto_ignored_modules:
            auto_ignored_modules.append(module)
```

Finnally, `auto_ignored_modules` will be considered as additional `ignored_modules` of FSDP. Is it a reasonable solution? If so, is it possible to implement this in the official implementation?

cc @zhaojuanmao @mrshenli @rohan-varma @awgu

**Comment by HAOCHENYE** [comment]
I made a mistake earlier. `ignored_modules` and `auto_wrap_policy` are closely related, and I can't deduce which modules should be `ignored` before finding the wrapped module.

BTW, why `ignored_parameters` and `ignored_modules` are not pass to sub-fsdp-module?

https://github.com/pytorch/pytorch/blob/f353d17755ed23b02924c962a86ff99a3405fe10/torch/distributed/fsdp/wrap.py#L393

It means that `ignored_parameters` and `ignored_modules` will not influence the sub-fsdp-module filtered by user-defined `auto_wrap_policy`

**Comment by awgu** [comment]
Thanks for filing this issue! Mixing frozen/non-frozen parameters is tricky for FSDP. There should not be any major downside to `use_orig_params=True` for general users, and we hope to make it default.

If FSDP ignores modules with mixed frozen/non-frozen parameters, how should users apply data parallelism to those parameters? Is the expectation to apply DDP?

---

> BTW, why `ignored_parameters` are not pass to sub-fsdp-module?
> 
> https://github.com/pytorch/pytorch/blob/f353d17755ed23b02924c962a86ff99a3405fe10/torch/distributed/fsdp/wrap.py#L393
> 
> It means that `ignored_parameters` will not influence the sub-fsdp-module filtered by user-defined `auto_wrap_policy`

The code pointer your linked is showing the actual wrapping via `_wrap()`, which does not recursively auto wrap. The root wraps on behalf of all of its submodules. I think the important thing is to pass `ignored_params` to `_recursive_wrap()` in the recursion:
https://github.com/pytorch/pytorch/blob/f353d17755ed23b02924c962a86ff99a3405fe10/torch/distributed/fsdp/wrap.py#L375-L382

**Comment by HAOCHENYE** [comment]
The reason I suggest passing `ignored_params` to `_wrap` is that sometimes the sub FSDP module specified by the `auto_wrap_policy` may have both frozen and non-frozen parameters. In this case, an error occurs when `use_orig_params` is set to False.

**Comment by awgu** [comment]
> for the first question, synchronizing the gradient for ignored_params and ignored_modules automatically makes sense to me.

Unfortunately, synchronizing gradients for ignored parameters is not the right behavior for everyone. That is semantic is more like "ignoring the parameter for flattening" rather than "ignoring the parameter". For example, there may be a model component is not data parallel (e.g. model-parallel embedding tables) that FSDP should ignore.

> The reason I suggest passing ignored_params to _wrap is that sometimes the sub FSDP module specified by the auto_wrap_policy may have both frozen and non-frozen parameters. In this case, an error occurs when use_orig_params is set to False.

I am not sure I follow this. The error is raised when one `FullyShardedDataParallel` instance with `use_orig_params=False` manages both frozen and non-frozen parameters. Are you saying that you are trying to ignore the frozen parameters or ignore the non-frozen parameters? Since FSDP will not do gradient reduction for ignored parameters (as mentioned above), this will be semantically incorrect.

---

If I am understanding correctly, the root issue is that we need better support for mixing frozen/non-frozen parameters. The existing constraint that each FSDP instance should be uniformly frozen or non-frozen is too restrictive. Is my understanding correct?
